server/src/main/java/org/opensearch/cluster/routing/RoutingNode.java

```diff
@@ -48,63 +48,140 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 /**
  * A {@link RoutingNode} represents a cluster node associated with a single {@link DiscoveryNode} including all shards
  * that are hosted on that nodes. Each {@link RoutingNode} has a unique node id that can be used to identify the node.
  */
 public class RoutingNode implements Iterable<ShardRouting> {
 
+    static class BucketedShards implements Iterable<ShardRouting> {
+        private static Map<Boolean, Integer> map = new HashMap<Boolean, Integer>() {
+            {
+                put(true, 0);
+                put(false, 1);
+            }
+        };
+
+        private final LinkedHashMap<ShardId, ShardRouting>[] shards; // LinkedHashMap to preserve order
+
+        BucketedShards(LinkedHashMap<ShardId, ShardRouting> primaryShards, LinkedHashMap<ShardId, ShardRouting> replicaShards) {
+            this.shards = new LinkedHashMap[2];
+            this.shards[0] = primaryShards;
+            this.shards[1] = replicaShards;
+        }
+
+        public boolean isEmpty() {
+            return this.shards[0].isEmpty() && this.shards[1].isEmpty();
+        }
+
+        public int size() {
+            return this.shards[0].size() + this.shards[1].size();
+        }
+
+        public boolean containsKey(ShardId shardId) {
+            return this.shards[0].containsKey(shardId) || this.shards[1].containsKey(shardId);
+        }
+
+        public ShardRouting get(ShardId shardId) {
+            if (this.shards[0].containsKey(shardId)) {
+                return this.shards[0].get(shardId);
+            }
+            return this.shards[1].get(shardId);
+        }
+
+        public ShardRouting add(ShardRouting shardRouting) {
+            return put(shardRouting.shardId(), shardRouting);
+        }
+
+        public ShardRouting put(ShardId shardId, ShardRouting shardRouting) {
+            ShardRouting ret = this.shards[map.get(shardRouting.primary())].put(shardId, shardRouting);
+            if (this.shards[map.get(!shardRouting.primary())].containsKey(shardId)) {
+                return this.shards[map.get(!shardRouting.primary())].remove(shardId);
+            }
+
+            return ret;
+        }
+
+        public ShardRouting remove(ShardId shardId) {
+            if (this.shards[0].containsKey(shardId)) {
+                return this.shards[0].remove(shardId);
+            }
+            return this.shards[1].remove(shardId);
+        }
+
+        @Override
+        public Iterator<ShardRouting> iterator() {
+            final Iterator<ShardRouting> iterator1 = Collections.unmodifiableCollection(shards[0].values()).iterator();
+            final Iterator<ShardRouting> iterator2 = Collections.unmodifiableCollection(shards[1].values()).iterator();
```

Comment:
nit: rename to primaryShardIterator/replicaIterator

Reply:
Done